### PR TITLE
feat(restrict-ingress): incorporate ingress-hostname-conflict

### PR DESCRIPTION
**Pull request by frazs**
Incorporates `ingress-hostname-conflict` into `restrict-hostnames` while expanding its functionality such that:
- Conflicts within a namespace are permitted
- Cross-namespace hostname conflicts are also caught between VirtualServices as well as across Ingresses and VirtualServices
- The `allowed-hosts` annotation can be used to permit any such cross-namespace conflicts for one or more particular hostpaths
- The namespaces and names of the conflicting resources are reported
  - Would there be any security concern with that? Someone who has the rights to create an Ingress or a VirtualService in any namespace could use this to discover the namespaces and names of Ingresses and VirtualServices corresponding to any hostname on that cluster. I am not sure whether that counts as sensitive information.
    - If so, it is trivial to change this to report only the namespace, or to only report that there is a conflict and suggest using a different hostname. When using a different hostname is not feasible, this could make knowing what is conflicting or who to contact to resolve the conflict more difficult, though that is likely an edge-case.
    
Mostly tested on InfraTest, but lacked time for the allowed-hosts label test (although that works in `opa test` and should be consistent with existing use of that function). Out of an abundance of caution, because I am new to Gatekeeper and about to leave for a week, on InfraTest I have not yet deleted the ingress-hostname-conflict template (which had earlier been modified to allow conflicts in the same namespace, but will not recognize allowed-hosts) or its constraints. 

This is my first policy and test, and I think I might still have some redundancies in the rego or test castes. Please feel free to suggest any improvements.

### CODE REVIEW

**Review by justbert — COMMENTED** [review]
Have you tested any failure scenarios?

What happens when someone attempts to create an exempted hostname but it would be a duplicate in another namespace?

Can you run `opa fmt`?

File: general/restrict-hostnames/rego/src.rego [hunk 1]
```diff
@@ -101,8 +97,55 @@ violation[{"msg": msg}] {
...
+	kind := input.review.kind.kind
+	re_match("^(Ingress|VirtualService)$", kind)
+	re_match("^(networking.k8s.io|networking.istio.io)$", input.review.kind.group)
+
+	hosts := {host | host := input.review.object.spec.rules[_].host} | {host | host := input.review.object.spec.hosts[_]}
+	host := hosts[_]
+	paths := {path | path := input.review.object.spec.rules[_].http.paths[_].path} | ({path | path := input.review.object.spec.http[_].match[_].uri.exact} | {path | path := input.review.object.spec.http[_].match[_].uri.prefix}) | {path | path := input.review.object.spec.http[_].match[_].uri.regex}
+	path := paths[_]
+
+	not is_allowed(host, path)
```
[the remark this answers is not in the document]
Comment (frazs) [reply general/restrict-hostnames/rego/src.rego:141]: For the failure scenarios, I had included [test_ingress_hostname_conflicts](https://github.com/StatCan/gatekeeper-policies/blob/425b6b847e9a191b2903fcd6abb99d01f0ec4d44/general/restrict-hostnames/rego/src_test.rego#L712) and [test_vs_hostname_conflicts](https://github.com/StatCan/gatekeeper-policies/blob/425b6b847e9a191b2903fcd6abb99d01f0ec4d44/general/restrict-hostnames/rego/src_test.rego#L994) which are functioning as expected on my end, reporting a conflict message for an exempted (but not allowed) hostname that exists in another namespace.

Thanks for informing me about `opa fmt`. I'll go ahead and apply it, as well as look into separating the redundant code into a function 😀

[the remark this answers is not in the document]
Comment (frazs) [reply general/restrict-hostnames/rego/src.rego]: I had some difficulty separating out the originally quoted code into a function. I attempted the following:
```
conflictable_hosts:= host {
	kind := input.review.kind.kind
	re_match("^(Ingress|VirtualService)$", kind)
	re_match("^(networking.k8s.io|networking.istio.io)$", input.review.kind.group)

	hosts := {host | host := input.review.object.spec.rules[_].host} | {host | host := input.review.object.spec.hosts[_]}
	host := hosts[_]
	paths := ({path | path := input.review.object.spec.rules[_].http.paths[_].path} | ({path | path := input.review.object.spec.http[_].match[_].uri.exact} | {path | path := input.review.object.spec.http[_].match[_].uri.prefix})) | {path | path := input.review.object.spec.http[_].match[_].uri.regex}
	path := paths[_]

	not is_allowed(host, path)
}
```
And referenced it with `host := conflictable_hosts`. While the first test case (conflicting Ingress) would pass, the second (conflicting VirtualService) would error out with `eval_conflict_error: complete rules must not produce multiple outputs`. I gathered that the workaround for this would be to make multiple definitions of `conflictable_hosts`, one checking only an Ingress and another only a VirtualService, but that seemed to add even more bulk... Though maybe something along those lines could be reused in the other violations.

That said while experimenting I found that it is possible to have [one hostname conflict violation instead of two](https://github.com/StatCan/gatekeeper-policies/blob/feat-restrict-ingress-frances/general/restrict-hostnames/rego/src.rego#L105). This is another approach to not duplicating the hostpath code, and results in a single message, for example `{{"msg": "Ingress hostname red.test.com conflicts with existing object(s) in other namespace(s): {\"Ingress/red/red_ingress\", \"VirtualService/red/red_vs\", \"VirtualService/red/red_vs2\"}"}}`.

I'll ponder whether I can make some further improvements.

[the remark this answers is not in the document]
Comment (frazs) [reply general/restrict-hostnames/rego/src.rego]: Per discussion in today's team meeting, we can go ahead with this for now, and take a crack at making the rego more elegant later on (perhaps as a Friday activity).

**Comment by justbert** [comment]
PS, LGTM!
